server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -750,43 +748,70 @@ protected void doRun() {
                         // the document listener gives us three-way logic: a document can fail processing (1), or it can
                         // be successfully processed. a successfully processed document can be kept (2) or dropped (3).
-                        final ActionListener<IngestPipelinesExecutionResult> documentListener = ActionListener.runAfter(
-                            new ActionListener<>() {
-                                @Override
-                                public void onResponse(IngestPipelinesExecutionResult result) {
-                                    assert result != null;
-                                    if (result.success) {
-                                        if (result.shouldKeep == false) {
-                                            onDropped.accept(slot);
-                                        } else {
-                                            assert firstPipeline != null;
-                                            firstPipeline.getMetrics().postIngestBytes(indexRequest.ramBytesUsed());
-                                        }
-                                    } else {
-                                        // We were given a failure result in the onResponse method, so we must store the failure
-                                        // Recover the original document state, track a failed ingest, and pass it along
-                                        updateIndexRequestMetadata(indexRequest, originalDocumentMetadata);
-                                        totalMetrics.ingestFailed();
-                                        onStoreFailure.apply(slot, result.failedIndex, result.exception);
+                        final ActionListener<Boolean> documentListener = ActionListener.runAfter(new ActionListener<>() {
+                            @Override
+                            public void onResponse(Boolean shouldKeep) {
+                                assert shouldKeep != null;
+                                if (shouldKeep == false) {
+                                    String originalIndex = indexRequest.index();
+                                    // We only increment the total counter for dropped docs here, because these docs don't reach the code
+                                    // that ordinarily take care of that.
+                                    // We reuse `shouldStoreFailure` here to determine whether the index request targets a data stream,
+                                    // because we only want to track these metrics for data streams.
+                                    Optional<Boolean> storeFailure = shouldStoreFailure.apply(originalIndex);
+                                    if (storeFailure.isPresent()) {
+                                        failureStoreMetrics.incrementTotal(originalIndex);
                                     }
```

Comment:
Reusing `shouldStoreFailure` here isn't ideal, but it avoids having to duplicate some logic. Maybe a method rename would already be sufficient?

Reply:
I see your concern, how about splitting it into two predicates or something and use the appropriate one each time?